### DISCUSSION TITLE
JMX MessageCount Inconsistency in Clustered Deployment

### DISCUSSION THREAD

**Discussion by jlplasce** [arkmq-org/arkmq-org-broker-operator#1029] [Q&A] [answered]
I have a JMX connection from a Java application using the "artemis-main" service to connect over 1099 in order to run a "getMessageCount()" against a queue. 
The initial connection seems fined, but when the messageCount query is ran, I get odd behavior.
The artemis-main service indeed shows me all 3 artemis pods
JMS traffic going over 62627 to that same artemis-main service are running just fine and working great

My main question is, do I have to connect to all 3 instances directly with JMX in order to get an accurate messageCount for a queue because of distribution? 

For context:
- Persistence Enabled
- MessageMigration Enabled
- Operator v1.2.4

For context on the Java implementation
```
JMXServiceURL jmxsu = JMXServiceURL("service:jm:rmi:///jndi/rmi://artemis-main.artemis-namespace:1099/jmxrmi");
JMXConnector connector = JMXConnectorFactory.connect(jmxsu);
MBeanServerConnection mbsc = connector.getMBeanServerConnection();

ObjectName on = ObjectNameBuilder.create(null, "Artemis").getQueueObjectName(SimpleString.toSimpleString("exampleQueue"), SimpleString.toSimpleString("exampleQueue"), RoutingType.ANYCAST);
QueueControl qc = MBeanServerInvocationHandler.newProxyInstance(mbsc, on, QueueControl.class, false);
Long count = qc.getMessageCount();

System.out.println(count.intValue());
```

**Comment by brusdev** [comment, marked as answer]
TL;DR you have to connect to all 3 instances directly with JMX in order to get an accurate messageCount for a queue because of distribution.

Each broker instance has its own MBean instances that only reflect the state of the broker instance that hosts them. When a JMX client query the message count of a queue it gets only the count of the messages in the queue of the broker instance to which it is connected.

**Reply by jlplasce** [reply]
This is what we thought was going to be the case. Thank you for the clarification

**Reply by brusdev** [reply]
Alternatively you could use a solution to aggregate metrics from all broker instances like Grafana to sum the count of the messages from all broker instances.
